File: frontend/src/utils/elementClassifier.ts
const JUMP_PATTERN = /^([1-4]?)(A|T|S|Lo|Lz|F)\b/;
const SPIN_PATTERN = /Sp[B0-4]?$/;
const STEP_PATTERN = /^(StSq|ChSq)/;
const LEVEL_PATTERN = /(\d)$/;

export type ElementType = "jump" | "spin" | "step";

export function classifyElement(name: string): ElementType | null {
  if (JUMP_PATTERN.test(name)) return "jump";
  if (SPIN_PATTERN.test(name)) return "spin";
  if (STEP_PATTERN.test(name)) return "step";
  return null;
}

export function isJumpElement(name: string): boolean {
  return JUMP_PATTERN.test(name);
}

export function isSpinElement(name: string): boolean {
  return SPIN_PATTERN.test(name);
}

export function isStepElement(name: string): boolean {
  return STEP_PATTERN.test(name);
}

export function extractJumpType(name: string): string | null {
  const m = name.match(JUMP_PATTERN);
  if (!m) return null;
  const rotation = m[1] || "1";
  return `${rotation}${m[2]}`;
}

export function elementLevel(name: string): number {
  const m = name.match(LEVEL_PATTERN);
  if (m) return parseInt(m[1], 10);
  if (/B$/.test(name)) return 0.5;
  return 0;
}

File: backend/app/services/element_classifier.py
"""Classify ISU element codes into jump, spin, or step categories."""

import re

_JUMP_PATTERN = re.compile(r"^([1-4]?)(A|T|S|Lo|Lz|F)\b")
_SPIN_PATTERN = re.compile(r"Sp[B0-4]?$")
_STEP_PATTERN = re.compile(r"^(StSq|ChSq)")
_LEVEL_PATTERN = re.compile(r"(\d)$")


def classify_element(name: str) -> str | None:
    """Classify an element code as 'jump', 'spin', 'step', or None."""
    if _JUMP_PATTERN.match(name):
        return "jump"
    if _SPIN_PATTERN.search(name):
        return "spin"
    if _STEP_PATTERN.match(name):
        return "step"
    return None


def extract_jump_type(name: str) -> str | None:
    """Extract the jump type with rotation count, e.g. '2A', '3Lz'."""
    m = _JUMP_PATTERN.match(name)
    if not m:
        return None
    rotation = m.group(1) or "1"
    jump_code = m.group(2)
    return f"{rotation}{jump_code}"


def extract_level(name: str) -> float:
    """Extract the level number from an element code. B suffix = 0.5, no level = 0."""
    m = _LEVEL_PATTERN.search(name)
    if m:
        return int(m.group(1))
    if name.endswith("B"):
        return 0.5
    return 0
